### PR TITLE
Happy api

### PR DESCRIPTION
**Pull request by elinaarenstrup**

### CODE REVIEW

File: server.js [hunk 1]
```diff
@@ -1,29 +1,95 @@
...
-app.use(bodyParser.json())
+app.use(cors());
+app.use(bodyParser.json());
 
 // Start defining your routes here
-app.get('/', (req, res) => {
-  res.send('Hello world')
-})
+app.get("/", (req, res) => {
+  res.send(listEndpoints(app));
```
Comment (HIPPIEKICK) [inline server.js:43]: ⭐

File: server.js [hunk 2]
```diff
@@ -1,29 +1,95 @@
...
+});
+
+//Get the 20 latest thoughts route, desc = newest first
+app.get("/thoughts", async (req, res) => {
+  const thoughts = await Thought.find()
+    .sort({ createdAt: "desc" })
+    .limit(20)
+    .exec();
+  res.json(thoughts);
```
Comment (HIPPIEKICK) [inline server.js:52]: Nice and clear!

File: server.js [hunk 3]
```diff
@@ -1,29 +1,95 @@
...
+app.post("/thoughts", async (req, res) => {
+  // Retrive the information sent by the client to our API endpoint
+  const { message } = req.body;
+  // Then use our mongoose model to create the database entry
+  const thought = new Thought({ message });
+
+  try {
+    // Success-case, send good status code to the client
+    const savedThought = await thought.save({ message });
+    res.status(200).json(savedThought);
```
Comment (HIPPIEKICK) [inline server.js:65]: Since you put the `message` in the `thought` on line 60, you don't need it on row 64 as well:
```
const thought = new Thought({ message });
  try {
    const savedThought = await thought.save();
    res.status(200).json(savedThought);
  }
```

File: server.js [hunk 4]
```diff
@@ -1,29 +1,95 @@
...
+  } catch (err) {
+    // Bad-req, send bad status code to the client
+    res.status(404).json({
```
Comment (HIPPIEKICK) [inline server.js:68]: Status 404 means "not found", 400 is "bad request" :)

File: server.js [hunk 5]
```diff
@@ -1,29 +1,95 @@
...
+      message: "Sorry, could not save this thought to database.",
+      errors: err.errors,
+    });
+  }
+});
+
+// Post likes to a specific message
+// Using POST and not PUT because I increment likes, not adding new data to database.
+app.post("/:id/like", async (req, res) => {
```
Comment (HIPPIEKICK) [inline server.js:77]: Since your other endpoints are named `/thoughts` - you could name this `/thoughts/:id/like` to be more consistent

File: server.js [hunk 6]
```diff
@@ -1,29 +1,95 @@
...
+  try {
+    const thought = await Thought.findOneAndUpdate(
+      { _id: req.params.id },
+      //increment by 1
+      { $inc: { hearts: 1 } }
+    );
```
Comment (HIPPIEKICK) [inline server.js:83]: 🥳

File: server.js [hunk 7]
```diff
@@ -1,29 +1,95 @@
...
+    res.json(thought).status(200);
```
Comment (HIPPIEKICK) [inline server.js:84]: Why did you change from 201? :)

**Comment by HIPPIEKICK** [comment]
I approved this, but I noticed a little late that the endpoints should look like below. Please take a look when you have time, and update it :)

### `GET /`

This endpoint should return a maximum of 20 thoughts, sorted by `createdAt` to show the most recent thoughts first.

### `POST /`

This endpoint expects a JSON body with the thought `message`, like this: `{ "message": "Express is great!" }` . If the input is valid (more on that below), the thought should be saved, and the response should include the saved thought object, including its `_id`.

### `POST /:thoughtId/like`

This endpoint doesn't require a JSON body. Given a valid thought id in the url, the API should find that thought, and update its `hearts` property to add one heart.

File: server.js (hunk 7 above)
Comment (elinaarenstrup) [reply server.js:84]: I got super confused when I saw that 200 was "ok" and 201 was "created", so I went with 200. Which one of these is the correct one to use? "Ok" sounded more correct to me but not sure.

Comment (HIPPIEKICK) [reply server.js:84]: Exactly, 201 is "created" (and it's the most common status code to use for when POST-requests are successful). I think 200 is good to use when you're unsure, since it's broader than e.g. 201, so neither are incorrect in this case! I was just curious of why you changed it :)
